meilisearch/tests/conftest.py

```diff
@@ -0,0 +1,59 @@
+# pylint: disable=redefined-outer-name
+import json
+from pytest import fixture
+
+from meilisearch.tests import common
+import meilisearch
+
+@fixture(scope='session')
+def client():
+    return meilisearch.Client(common.BASE_URL, common.MASTER_KEY)
+
+@fixture(autouse=True)
+def clear_indexes(client):
+    """
+    Auto clear the indexes after each tetst function run.
+    helps with identifying tests that are not independent.
+    """
+    # Yield back to the test function
+    yield
+    # test function has finished, let's cleanup
+    indexes = client.get_indexes()
+    for index in indexes:
+        client.index(index['uid']).delete()
+
+
+@fixture(scope='function')
+def indexes_sample(client):
+    indexes = []
+    for index_args in common.INDEX_FIXTURE:
+        indexes.append(client.create_index(**index_args))
+    # yield the indexes to the test so it can use it
+    yield indexes
+
+
+@fixture(scope='session')
+def small_movies():
+    """
+    Run once per session, provide the content of small_movies.json
+     as a dictionary to the test.
+    """
+    with open('./datasets/small_movies.json', 'r') as movie_file:
+        yield json.loads(movie_file.read())
+
+
+@fixture(scope='function')
+def empty_index(client):
+    def index_maker(index_name=common.INDEX_UID):
+        return client.create_index(uid=index_name)
+    return index_maker
```

Comment:
Sorry not to have asked before, but with my colleague @eskombro we were asking what is the purpose of doing a function in another function? 🙂

Reply:
it is creating higher order function, you predefine some things and pass some other things later on.

in pytest-context:
fixtures already get called before a test runs, and test gets the result.
But what if we want a fixture that does something slightly differently depending on a param our tests passes?
We can define a fixture(empty_index), that returns a function to make indexes.

In this specific case, `index_name` by default has value, but maybe a function uses this multiple times to create many indexes with slightly different names: `index_maker("index-1") then index_maker("index-2")`

It can be avoided of course, test can import a simple helper function, that function imports `(meili) client` and depending on parameters that the test has given it creates an index.
But this approach can be applied for any fixture, every test can call helper functions to achieve stuff instead of fixture, just wanted to stick to the fixture way of doing it.

function-in-function here can be I think thought of very similar(even same) thing to python decorators do or javascript HOCs.

Please let me know if I was ambiguous or missed the point :D

Reply:
Got it!! Thanks!